compiler/rustc_hir/src/hir.rs

```diff
@@ -3279,6 +3284,16 @@ impl<'hir> AssocItemConstraintKind<'hir> {
     }
 }
 
+#[derive(Debug, Clone, Copy, PartialEq, HashStable_Generic)]
+pub enum TySource {
+    /// `Vec` in `Vec::new`
+    ImplicitSelf,
+
+    /// Details not yet needed. Feel free to give useful
+    /// categorization to these usages.
+    Other,
+}
```

Comment:
Can we avoid having this enum and have the lint compute where the type appears?

Reply:
Specifically: you can look at the parent, or implement the lint on paths and look at the first segment